text/0000-partial_types.md

```diff
@@ -0,0 +1,642 @@
+
+`<%permit>` - field is accessible (always implicit), `%deny` - field is forbidden to use, `%miss` - it is uninitialized field, field is forbidden to use.
+
+Let we wish to have next structure:
+```rust
+struct SR <T>{
+    val : T,
+    lnk : & T, // reference to val
+}
+```
+Now it is impossible to write immutable self-referential types. But with partial initialization it becomes easy:
+```rust
+let x = SR {val : 5i32 };
+    // omitted field is uninitialized, it has %miss field-access
+    // x : SR<i32>.{val, %miss lnk}
+
+x.lnk let= & x.val;
+    // (let=) "extender", late initialized field
+    // x : SR<i32>;
+    // x : SR<i32>.%full;
+```
+Easy-peasy! We create partial type and extend by late initializing fields!
+
+```rust
+let baz = (7i16, &51i32, "some_other_string").{1, 2, %unfill};
+    // or set access-filer before tuple constructor 
+    // baz : (i16, &i32, &str).{1, 2, %miss 0};
+
+let bar = (%miss 7i16, &51i32, "some_other_string");
+    // or set access-filer before tuple constructor 
+    // baz : (i16, &i32, &str).{1, 2, %miss 0};
+
+bar.0 let= 68;
+	// (let=) "extender", late initialized field
+	// bar : (i16, &i32, &str);
+	// bar : (i16, &i32, &str).%full;
+```
+Same with tuples.
+
+Ok, let's try a bit more impossible now - two mutual borrowings of same variable (but different fields):
+```rust
+struct Point {
+  x: f64,
+  y: f64, 
+  was_x: f64, 
+  was_y: f64,
+  state : f64,
+}
+let mut p1 = Point {x:1.0, y:2.0, was_x: 4.0, was_y: 5.0, state: 12.0};
+	// p1 : Point
+	
+let ref_p1was = &mut p1.{wax_x, was_y};
+	// ".{}" is a getter several fields, similar to "." get 1 field 
+	// pfull : &mut Point.%{was_x, was_y, %unfill}
+	
+let ref_p1now = &mut p1.{x, y};
+	// pfull : &mut Point.%{x, y, %unfill}
+	
+let Point {%deny x, %deny y, was_x: upd_x, was_y: upd_y, %deny state} = ref_p1was;
+    // all %deny and %miss fields must be '_', error in other case
+	// similar with binding deconstruction
+```
+It is simple and will be possible. 
+
+Same easy to write functions, which consume partial parameters:
+```rust
+fn re_ref_t (&self : & Self.%{t, %any}) -> &f64 {
+   &self.t
+}
+
+fn refmut_w (&mut self : &mut Self.%{w, %any}) -> &mut f64 {
+   &mut self.w
+}
+```
+
+We could do much harder things: we wish parallel using function implementation, which update either `x` or `y` and both read `state` field.
+```rust
+impl {
+	pub fn mx_rstate(&mut self1 : &mut Self.%{x, %any}, &self2 : & Self.{state, %any})
+	{ /* ... */ }
+		
+	pub fn my_rstate(&mut self1 : &mut Self.%{y, %any}, &self2 : & Self.{state, %any})  
+	{ /* ... */ }
+
+	// Unfortunately it is a mandatory to extensive mut-borrow 'state' field!
+	pub fn mxystate(&mut self : &mut Self.%{x, y, state, %any})  
+	{ 
+		/* ... */
+		self.{x, state}.mx_rstate();
+		/* ... */
+		self.{y, state}.my_rstate();
+		/* ... */		
+	}
+}
+```
+We need for this ... several selfs! Wow!
```

Comment:
`Add` is a bit magical, but it is a interesting idea!

It is good looking for trait implementations, but it is a unclear for inherit  implementations.
Furthermore `Rhs` is right-hand-side for 2 variable. But we have one variable. So,  maybe general parameter `Smv` as "same variable" or `Sngl` as "single variable" is a better choice.
```rust
impl Point<Smv = Self> { /* */ }
```

Reply:
I didn't realize that both `self1` and `self2` refer to the same value. That is very surprising, because it looks like the function has two arguments, but it has only one. I don't think this is a good idea.

Reply:
Why does this need to be written as 
```rust
pub fn mx_rstate(self1: &mut Self.{x, %any}, self2: &Self.{state, %any})
```
instead of
```rust
pub fn mx_rstate(self: &mut Self.{x, state, %any})
```
?
If the reason is to represent that x is not borrowed mutably, you could do that with a syntax like `self: Self.{ref x, ref mut state}`.

Reply:
I don't think that would work, because then `self: Self.{ref x, state}` would mean that `self.x` is passed by reference and `self.state` by value. But these annotations should just inform the borrow checker what is borrowed and how, they shouldn't change program semantics.

Reply:
well that could just be an error, or you could simplify the syntax to omit the `ref`.

Actually it is not unreasonable to pass `self.state` by value, this sort of thing goes under the name of "`&move` references" and it would be useful for a lot of things.

Reply:
@digama0  This is the whole point of partial borrowing ))  We wish `state`-field has parallel reference of another function

It is 2 alternatives - several selfs or partial mutability:
```rust
// several selfs (or Smv parameter)
pub fn mx_rstate(self1: &mut Self.{x, %any}, self2: &Self.{state, %any}) { /* */ }

// or partial mutability
pub fn mx_rstate(self: &mut.{x} Self.{x, state, %any}) { /* */ }
```

> If the reason is to represent that x is not borrowed mutably, you could do that with a syntax like `self: Self.{ref x, ref mut state}`

Yes, this creates ambiguity.
```rust
let r1 = &mut p.{x, y};  // is it ~ r2(all mutable) or ~r4 (all immutable)
let r2 = &mut p.{mut x, mut y};
let r3 = &mut p.{mut x, y};

let r4 = & p.{x, y};
let r5 = & p.{mut x, mut y}; // is it ~ r2(all mutable), is it valid at all
let r6 = & p.{mut x, y}; // is it valid at all

let p1 = p.{x, y};
let p2 = p.{mut x, mut y};  // is it ~ p5(all mutable), is it valid at all
let p3 = p.{mut x, y}; // is it valid at all

let mut p4 = p.{x, y};  // is it ~ p5(all mutable) or ~p1 (all immutable)
let mut p5 = p.{mut x, mut y};
let mut p6 = p.{mut x, y};

fn cutvr<T> (var : T) {
 ... var.{x,y} ... // what to do here? is it var a ref? mutref?  notref?
}
```

Reply:
I don't think that is an ambiguity so much as having lots of overlapping and redundant syntax, and needing to make syntax decisions (which likely won't matter since this syntax will be revised another 10 times before reaching some kind of fixed point).

If you are asking what I would do about that syntax, I would say that you can't write `&mut p.{...}` at all. If you use the `.{...}` syntax then you have to put the muts in there. Leave future "partial types ergonomics" stuff for the future.

I would actually say that all of r1-r6, p1-p6 are an error, because you can't write partial types in expression position. Instead, you can have them in type position:
```rust
struct S { x: u8, y: u8 };
let r1: &mut S.{x, y};  // error, suggest &S.{mut x, mut y} instead
let r2: &mut S.{mut x, mut y};  // error, suggest &S.{mut x, mut y} instead
let r3: &mut S.{mut x, y};  // error, suggest &S.{mut x, mut y} instead

let r4: &S.{x, y}; // same as `r4: &S`
let r5: &S.{mut x, mut y}; // same as `r5: &mut S
let r6: &S.{mut x, y}; // partial mutable reference

let p1: S.{x, y}; // same as `p1: S`
let p2: S.{mut x, mut y};  // error, mut not associated to a pointer type
let p3: S.{mut x, y};  // error, mut not associated to a pointer type

let q1.{x, y}: S; // same as `q1: S`
let q2.{mut x, mut y}: S;  // same as `mut q2: S`
let q3.{mut x, y}: S;  // q3.x is a mutable binding, q3.y is not

let mut q4.{x, y}: S;  // error, suggest `q4.{mut x, mut y}: S` instead
let mut q5.{mut x, mut y}: S;   // error, suggest `q5.{mut x, mut y}: S` instead
let mut q6.{mut x, y}: S;  // error, suggest `q6.{mut x, mut y}: S` instead

fn cutvr<T> (var : T) {
 ... var.{x,y} ... // not legal, both because T doesn't have an x field
                   // and because var is an expression, not a type or pattern
 }
 ```